File: leetcodeSolutions.cs
public class Solution
{
    // Merge Alternatively
    // merges two strings alternatively. say you habe abcd and efgh, the outcome will be aebfcgdh and it will append the rest of the string to the back it they are not of the same length
    public string MergeAlternately(string word1, string word2)
    {
        bool sameLength = word1.Length == word2.Length;
        string longWord = (word1.Length > word2.Length) ? word1 : word2;
        string shortWord = (word1 == longWord) ? word2 : word1;

        string returnWord = "";
        for (int i = 0; i < shortWord.Length; i++)
        {
            returnWord += word1[i];
            returnWord += word2[i];
        }
        if (!sameLength)
        {
            returnWord += longWord[shortWord.Length..];
        }
        return returnWord;
    }

    // Greatest Common Divisor of Strings
    // This one takes two strings and finds out the longest string that can perrfectly devided the two. if you have abcd and abcdabcd, the ooutcome will be abcd. it will return and empty string if there is no common devisor
    public string GcdOfStrings(string str1, string str2)
    {
        int str1Length = str1.Length;
        int str2Length = str2.Length;

        int shortLength = (str1Length > str2Length) ? str2Length : str1Length;
        var str1Solutions = new List<string>();

        for (int i = 0; i < shortLength; i++)
        {
            if (!((str1Length % (i + 1) == 0) && (str2Length % (i + 1) == 0)))
            {
                continue;
            }
            string multipliedString = string.Concat(Enumerable.Repeat(str1[..(i + 1)], str1Length / (i + 1)));
            if (str1 == multipliedString)
            {
                str1Solutions.Add(str1[..(i + 1)]);
            }
        }

        for (int i = str1Solutions.Count - 1; i >= 0; i--)
        {
            string multipliedString = string.Concat(Enumerable.Repeat(str1Solutions[i], str2Length / str1Solutions[i].Length));

            if (str2 == multipliedString) { return str1Solutions[i]; }
        }
        return "";

    }
    public IList<bool> KidsWithCandies(int[] candies, int extraCandies)
    {
        List<bool> result = [];
        int maxCandy = candies.Max();
        foreach (int candy in candies)
        {
            result.Add(candy + extraCandies >= maxCandy);
        }

        return result;
    }
    public bool CanPlaceFlowers(int[] flowerbed, int n)
    {
        for (int i = 0; i < flowerbed.Length; i++)
        {
            if (n <= 0) break;
            if (flowerbed.Length == 1)
            {
                if (flowerbed[i] == 0)
                {
                    flowerbed[i] = 1;
                    n--;
                }
                continue;
            }
            if (i == 0)
            {
                if (flowerbed[i] == 0 && flowerbed[i + 1] == 0)
                {
                    flowerbed[i] = 1;
                    i++;
                    n--;
                }
                continue;
            }
            if (i == (flowerbed.Length - 1))
            {
                if (flowerbed[i] == 0 && flowerbed[i - 1] == 0)
                {
                    flowerbed[i] = 1;
                    i++;
                    n--;
                }
                continue;
            }
            if (flowerbed[i] == 0 && flowerbed[i - 1] == 0 && flowerbed[i + 1] == 0)
            {
                flowerbed[i] = 1;
                i++;
                n--;
                continue;
            }
            if (flowerbed[i] == 1) { i++; }
        }
        if (n <= 0) { return true; }
        return false;
    }

    public string ReverseVowels(string s)
    {
        char[] vowels = [
        'a',
        'e',
        'i',
        'o',
        'u',
        'A',
        'E',
        'I',
        'O',
        'U',
        ]
        ;
        List<char> sVowels = [];
        List<int> index = [];
        char[] stringlist = s.ToArray();

        for (int i = 0; i < stringlist.Length; i++)
        {
            char letter = (char)stringlist[i];
            if (vowels.Contains(letter))
            {

                sVowels.Add(letter);
                index.Add(i);
            }
        }
        sVowels.Reverse();
        for (int i = 0; i < sVowels.Count; i++)
        {
            char replaceVowel = sVowels[i];
            stringlist[index[i]] = replaceVowel;
        }
        return string.Join(string.Empty, stringlist);

    }
    public string ReverseWords(string s)
    {
        string[] wordList = s.Split(" ");
        string newWord = "";
        for (int i = wordList.Length - 1; i >= 0; i--)
        {
            if (wordList[i] != string.Empty)
            {
                newWord += wordList[i] + " ";
            }
        }
        return newWord.Trim();

    }
    public int[] ProductExceptSelf(int[] nums)
    {
        int[] prefix = Enumerable.Repeat(1, nums.Length).ToArray();
        int[] suffix = Enumerable.Repeat(1, nums.Length).ToArray();
        int prefixNum = 1;
        int suffixNum = 1;
        for (int i = 0; i < nums.Length; i++)
        {
            int j = nums.Length - 1 - i;
            if (i == 0) { continue; }

            prefixNum *= nums[i - 1];
            prefix[i] = prefixNum;

            suffixNum *= nums[j + 1];
            suffix[j] = suffixNum;
        }
        for (int i = 0; i < nums.Length; i++)
        {
            nums[i] = prefix[i] * suffix[i];
        }
        return nums;
    }
    public bool IncreasingTriplet(int[] nums)
    {
        int[] lowPair = [0, 0];
        int leastNum = nums[0];
        int secondLowest;
        bool setLowPair = false;

        for (int i = 0; i < nums.Length; i++)
        {
            if (i == 0) { continue; }
            if (setLowPair && nums[i] > lowPair[1]) { return true; }
            if (nums[i] > leastNum)
            {
                secondLowest = nums[i];
                if (setLowPair && secondLowest < lowPair[1])
                {
                    lowPair = [leastNum, secondLowest];
                    setLowPair = true;
                }
                if (!setLowPair)
                {
                    lowPair = [leastNum, secondLowest];
                    setLowPair = true;
                }
            }
            if (nums[i] < leastNum)
            {
                leastNum = nums[i];
            }
        }
        return false;
    }
    public int Compress(char[] chars)
    {
        string s = "";
        int charLength = 1;
        for (int i = 0; i < chars.Length; i++)
        {
            if (i == 0)
            {
                s += chars[i];
                continue;
            }

            if (i == chars.Length - 1 && chars[i] == chars[i - 1])
            {
                s += charLength + 1;
            }
            if (chars[i] == chars[i - 1])
            {
                charLength++; continue;
            }
            if (charLength > 1)
            {
                s += charLength;
            }
            s += chars[i];
            charLength = 1;

        }
        s.CopyTo(0, chars, 0, s.Length);
        chars = s.ToArray();
        return chars.Length;
    }
    public void MoveZeroes(int[] nums)
    {
        int lastSwapPos = -1;
        for (int i = 0; i < nums.Length; i++)
        {
            if (nums[i] == 0) { continue; }
            if (lastSwapPos + 1 == i) { lastSwapPos++; continue; }
            // swap
            nums[lastSwapPos + 1] = nums[i];
            nums[i] = 0;
            lastSwapPos++;

        }
    }

    public bool IsSubsequence(string s, string t)
    {
        int sIndex = 0;
        if (s == "") { return true; }
        for (int i = 0; i < t.Length; i++)
        {
            if (t[i] == s[sIndex] && sIndex >= s.Length - 1) { return true; }
            if (t[i] == s[sIndex]) { sIndex++; }
        }
        return false;

    }

    public int MaxArea(int[] height)
    {
        int maxArea = 0;
        for (int i = 0, j = height.Length - 1; i >= j;)
        {
            int area = (j - i) * ((height[i] < height[j]) ? height[i] : height[j]);
            if (area > maxArea) { maxArea = area; }
            if (height[i] <= height[j])
            {
                i++;
            }
            else
            {
                j--;
            }

            Console.WriteLine(area);

        }
        return maxArea;

    }
    public int MaxOperations(int[] nums, int k)
    {
        int numOfOperations = 0;
        var secondPairs = new Dictionary<int, int>();

        for (int i = 0; i < nums.Length; i++)
        {
            int value;
            bool numInDic = secondPairs.TryGetValue(nums[i], out value);
            if (numInDic && value > 0) { secondPairs[nums[i]] = value - 1; numOfOperations++; }
            else if (secondPairs.TryGetValue(k - nums[i], out value)) { secondPairs[k - nums[i]] = value + 1; }
            else { secondPairs[k - nums[i]] = 1; }

        }
        return numOfOperations;
    }
    public double FindMaxAverage(int[] nums, int k)
    {
        int sum = 0;
        for (int i = 0; i < k; i++)
        {
            sum += nums[i];
        }
        int maxSum = sum;
        for (int i = k; i < nums.Length; i++)
        {
            sum -= nums[i - k];
            sum += nums[i];
            if (sum > maxSum)
            {
                maxSum = sum;
            }
        }
        return maxSum / (double)k;
    }


    public int MaxVowels(string s, int k)
    {
        char[] vowels = ['a', 'e', 'i', 'o', 'u'];
        int num = 0;
        for (int i = 0; i < k; i++)
        {
            if (vowels.Contains(s[i])) { num++; }
        }
        int maxVowel = num;
        for (int i = k; i < s.Length; i++)
        {
            if (num == k) { return num; }
            bool vIn = vowels.Contains(s[i]);
            bool vOut = vowels.Contains(s[i - k]);
            if (vIn && !vOut) { num++; }
            if (!vIn && vOut) { num--; }
            if (num > maxVowel) { maxVowel = num; }
        }
        return maxVowel;
    }


    public int LongestOnes(int[] nums, int k)
    {
        int zereos = 0;
        int i = 0;
        int j = 0;
        while (j < nums.Length)
        {
            if (nums[j] == 0) { zereos++; }
            j++;

            if (zereos > k)
            {
                if (nums[i] == 0) { zereos--; }
                i++;
            }
        }
        return j - i;


    }

    public int LongestSubarray(int[] nums)
    {
        int longest = 0;
        int lastCount = 0;
        int count = 0;
        bool encounteredZero = false;
        for (int i = 0; i < nums.Length; i++)
        {
            if (nums[i] == 0)
            {
                encounteredZero = true;
                if (count + lastCount > longest) { longest = count + lastCount; }
                lastCount = count;
                count = 0;
            }
            else
            {
                count++;
            }

            if (i == nums.Length - 1)
            {
                if (count + lastCount > longest) { longest = count + lastCount; }
            }
        }
        if (!encounteredZero) { return longest - 1; }
        return longest;
    }

    public int LargestAltitude(int[] gain)
    {
        int highest = 0;
        int current = 0;
        foreach (int g in gain)
        {
            current += g;
            if (current > highest)
            {
                highest = current;
            }
        }
        return highest;
    }

    public int PivotIndex(int[] nums)
    {
        int[] prefixArr = new int[nums.Length];
        int[] suffixArr = new int[nums.Length];
        int prefix = 0;
        int suffix = 0;
        for (int i = 0, j = nums.Length - 1; i < nums.Length && j >= 0; i++, j--)
        {
            prefixArr[i] = prefix;
            suffixArr[j] = suffix;
            prefix += nums[i];
            suffix += nums[j];


        }
        for (int i = 0; i < nums.Length; i++)
        {
            if (prefixArr[i] == suffixArr[i]) { return i; }

        }
        return -1;
    }

    public IList<IList<int>> FindDifference(int[] nums1, int[] nums2)
    {
        var num1Set = nums1.ToHashSet<int>();
        var num2Set = nums2.ToHashSet<int>();
        foreach (int num in num2Set)
        {
            if (num1Set.Contains(num))
            {
                num1Set.Remove(num);
                num2Set.Remove(num);
            }
        }

        var num1Distinct = num1Set.ToArray<int>();
        var num2Distinct = num2Set.ToArray<int>();
        return [num1Distinct, num2Distinct];
    }
    public bool UniqueOccurrences(int[] arr)
    {
        var numObj = new Dictionary<int, int>();
        var occObj = new Dictionary<int, int>();
        foreach (int num in arr)
        {
            if (numObj.TryGetValue(num, out int value))
            {
                numObj[num]++;
            }
            else { numObj[num] = 1; }
        }
        foreach (var pair in numObj)
        {
            if (occObj.TryGetValue(pair.Value, out int value))
            {
                return false;
            }
            else { occObj[pair.Value] = 1; }
        }
        return true;
    }

    public bool CloseStrings(string word1, string word2)
    {
        if (word1.Length != word2.Length) { return false; }
        for (int i = 0; i < word1.Length; i++)
        {
            if (!word1.Contains(word2[i]))
            {
                return false;
            }
        }

        var word1Obj = new Dictionary<char, int>();
        var word2Obj = new Dictionary<char, int>();
        for (int i = 0; i < word1.Length; i++)
        {
            if (word1Obj.TryGetValue(word1[i], out int occ1))
            {
                word1Obj[word1[i]]++;
            }
            else { word1Obj[word1[i]] = 1; }
            if (word2Obj.TryGetValue(word2[i], out int occ2))
            {
                word2Obj[word2[i]]++;
            }
            else { word2Obj[word2[i]] = 1; }
        }

        var word1Array = word1Obj.ToList();
        var word2Array = word2Obj.ToList();

        if (word1Array.Count != word2Array.Count)
        {
            return false;
        }

        List<int> word1List = [];
        List<int> word2List = [];
        for (int i = 0; i < word1Array.Count; i++)
        {
            word1List.Add(word1Array[i].Value);
            word2List.Add(word2Array[i].Value);
        }
        word1List.Sort();
        word2List.Sort();
        for (int i = 0; i < word1List.Count; i++)
        {
            if (word1List[1] != word2List[1]) { return false; }
        }

        return true;
    }

    public string RemoveStars(string s)
    {

        List<char> answer = [];
        for (int i = 0; i < s.Length; i++)
        {
            if (s[i] == '*')
            {
                answer.RemoveAt(answer.Count - 1);
                continue;
            }
            answer.Add(s[i]);
        }
        return string.Join(string.Empty, answer);
    }

    public int EqualPairs(int[][] grid)
    {
        int number = 0;
        var rows = new Dictionary<string, int>();
        var columns = new List<string>();

        for (int i = 0; i < grid.Length; i++)
        {
            string row = "";
            string column = "";
            for (int j = 0; j < grid.Length; j++)
            {
                row += $"-{grid[i][j]}";
                column += $"-{grid[j][i]}";
            }
            if (rows.ContainsKey(row))
            {
                rows[row] += 1;
            }
            else { rows[row] = 1; }

            columns.Add(column);

        }

        for (int i = 0; i < columns.Count; i++)
        {
            if (rows.ContainsKey(columns[i]))
            {
                number += rows[columns[i]];
            }
        }
        return number;


    }
}

File: leetcodeSolutions.ts
function mergeAlternately(
    word1: string,
    word2: string
): string {
    const sameLength = word1.length === word2.length
    const longWord =
        word1.length > word2.length ? word1 : word2
    const shortWord = word1 == longWord ? word2 : word1

    let returnWord = ""
    for (let i = 0; i < shortWord.length; i++) {
        returnWord += word1[i]
        returnWord += word2[i]
    }
    if (!sameLength) {
        returnWord += longWord.slice(shortWord.length)
    }
    return returnWord
}

function gcdOfStrings(str1: string, str2: string): string {
    const str1Length = str1.length
    const str2Length = str2.length

    const shortLength =
        str1Length > str2Length ? str2Length : str1Length
    const str1Solutions: string[] = []

    for (let i = 0; i < shortLength; i++) {
        if (
            !(
                str1Length % (i + 1) == 0 &&
                str2Length % (i + 1) == 0
            )
        ) {
            continue
        }
        let multipliedString = str1
            .slice(0, i + 1)
            .repeat(str1Length / (i + 1))
        if (str1 == multipliedString) {
            str1Solutions.push(str1.slice(0, i + 1))
        }
    }
    for (let i = str1Solutions.length - 1; i >= 0; i--) {
        let multipliedString = str1Solutions[i].repeat(
            str2Length / str1Solutions[i].length
        )

        if (str2 == multipliedString) {
            return str1Solutions[i]
        }
    }

    return ""
}

function kidsWithCandies(
    candies: number[],
    extraCandies: number
): boolean[] {
    const result: boolean[] = []
    const maxCandy = Math.max(...candies)
    for (const candy of candies) {
        result.push(candy + extraCandies >= maxCandy)
    }

    return result
}

function canPlaceFlowers(
    flowerbed: number[],
    n: number
): boolean {
    for (let i = 0; i < flowerbed.length; i++) {
        if (n <= 0) break
        if (flowerbed.length == 1) {
            if (flowerbed[i] == 0) {
                flowerbed[i] = 1
                n--
            }
            continue
        }
        if (i == 0) {
            if (
                flowerbed[i] == 0 &&
                flowerbed[i + 1] == 0
            ) {
                flowerbed[i] = 1
                i++
                n--
            }
            continue
        }
        if (i == flowerbed.length - 1) {
            if (
                flowerbed[i] == 0 &&
                flowerbed[i - 1] == 0
            ) {
                flowerbed[i] = 1
                i++
                n--
            }
            continue
        }
        if (
            flowerbed[i] == 0 &&
            flowerbed[i - 1] == 0 &&
            flowerbed[i + 1] == 0
        ) {
            flowerbed[i] = 1
            i++
            n--
            continue
        }
        if (flowerbed[i] == 1) {
            i++
        }
    }
    if (n <= 0) {
        return true
    }
    return false
}

function reverseVowels(s: string): string {
    const vowels = [
        "a",
        "e",
        "i",
        "o",
        "u",
        "A",
        "E",
        "I",
        "O",
        "U",
    ]
    const sVowels: string[] = []
    const index: number[] = []
    const stringlist = s.split("")

    for (let i = 0; i < stringlist.length; i++) {
        const letter = stringlist[i]
        if (vowels.includes(letter)) {
            sVowels.push(letter)
            index.push(i)
        }
    }
    sVowels.reverse()
    for (let i = 0; i < sVowels.length; i++) {
        const replaceVowel = sVowels[i]
        stringlist[index[i]] = replaceVowel
    }
    return stringlist.join("")
}

function reverseWords(s: string): string {
    const wordList = s.split(" ")
    let newWord = ""
    for (let i = wordList.length - 1; i >= 0; i--) {
        if (!wordList[i]) {
            newWord += wordList[i] + " "
        }
    }
    return newWord.trim()
}

function productExceptSelf(nums: number[]): number[] {
    const array: number[][] = Array(nums.length)
        .fill(null)
        .map(() => [1, 1])
    let prefixNum = 1
    let suffixNum = 1
    for (let i = 0; i < nums.length; i++) {
        let j = nums.length - 1 - i
        if (i == 0) {
            continue
        }

        prefixNum *= nums[i - 1]
        array[i][0] = prefixNum

        suffixNum *= nums[j + 1]
        array[j][1] = suffixNum
    }

    for (let i = 0; i < nums.length; i++) {
        nums[i] = array[i][0] * array[i][1]
    }
    return nums
}

function increasingTriplet(nums: number[]): boolean {
    let lowPair = [0, 0]
    let leastNum = nums[0]
    let secondLowest
    let setLowPair = false

    for (let i = 0; i < nums.length; i++) {
        if (i == 0) {
            continue
        }
        if (setLowPair && nums[i] > lowPair[1]) {
            return true
        }
        if (nums[i] > leastNum) {
            secondLowest = nums[i]
            if (setLowPair && secondLowest < lowPair[1]) {
                lowPair = [leastNum, secondLowest]
                setLowPair = true
            }
            if (!setLowPair) {
                lowPair = [leastNum, secondLowest]
                setLowPair = true
            }
        }
        if (nums[i] < leastNum) {
            leastNum = nums[i]
        }
    }
    return false
}
function compress(chars: string[]): number {
    let s = ""
    let charLength = 1
    for (let i = 0; i < chars.length; i++) {
        if (i == 0) {
            s += chars[i]
            continue
        }

        if (
            i == chars.length - 1 &&
            chars[i] == chars[i - 1]
        ) {
            s += charLength + 1
        }
        if (chars[i] == chars[i - 1]) {
            charLength++
            continue
        }
        if (charLength > 1) {
            s += charLength
        }
        s += chars[i]
        charLength = 1
    }

    const newList = [...s.split("")]
    chars.splice(0, s.length, ...newList)
    chars = newList
    return chars.length
}

function moveZeroes(nums: number[]): void {
    let lastSwapPos = -1
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] === 0) {
            continue
        }
        if (lastSwapPos + 1 === i) {
            lastSwapPos++
            continue
        }
        nums[lastSwapPos + 1] = nums[i]
        nums[i] = 0
        lastSwapPos++
    }
}
function isSubsequence(s: string, t: string): boolean {
    let sIndex = 0
    if (s == "") {
        return true
    }
    for (let i = 0; i < t.length; i++) {
        if (t[i] == s[sIndex] && sIndex >= s.length - 1) {
            return true
        }
        if (t[i] == s[sIndex]) {
            sIndex++
        }
    }
    return false
}

function maxArea(height: number[]): number {
    let maxArea = 0
    for (let i = 0, j = height.length - 1; i <= j; ) {
        let area =
            (j - i) *
            (height[i] < height[j] ? height[i] : height[j])
        if (area > maxArea) {
            maxArea = area
        }
        if (height[i] <= height[j]) {
            i++
        } else {
            j--
        }
    }
    return maxArea
}

function maxOperations(nums: number[], k: number): number {
    let ops = 0
    let set = {}

    for (let num of nums) {
        const pair = k - num
        if (set[num] >= 1) {
            set[num]--
            ops++
        } else {
            set[pair] ? set[pair]++ : (set[pair] = 1)
        }
    }
    return ops
}

function findMaxAverage(nums: number[], k: number): number {
    let maxSum = 0
    let sum = 0
    for (let i = 0; i < k; i++) {
        sum += nums[i]
    }
    maxSum = sum
    for (let i = k; i < nums.length; i++) {
        sum -= nums[i - k]
        sum += nums[i]
        if (sum > maxSum) {
            maxSum = sum
        }
    }
    return maxSum / k
}

function maxVowels(s: string, k: number): number {
    const vowels = ["a", "e", "i", "o", "u"]
    let num = 0
    for (let i = 0; i < k; i++) {
        if (vowels.includes(s[i])) {
            num++
        }
    }
    let maxVowel = num
    for (let i = k; i < s.length; i++) {
        if (num == k) {
            return num
        }
        let vIn = vowels.includes(s[i])
        let vOut = vowels.includes(s[i - k])
        if (vIn && !vOut) {
            num++
        }
        if (!vIn && vOut) {
            num--
        }
        if (num > maxVowel) {
            maxVowel = num
        }
    }
    return maxVowel
}

function longestOnes(nums: number[], k: number): number {
    if (nums === null) {
        throw new Error("Input array is null")
    }

    let start = 0,
        end = 0,
        zeros = 0

    while (end < nums.length) {
        if (nums[end] === 0) {
            zeros++
        }
        end++

        if (zeros > k) {
            if (nums[start] === 0) {
                zeros--
            }
            start++
        }
    }

    return end - start
}

function longestSubarray(nums: number[]): number {
    let longest = 0
    let lastCount = 0
    let count = 0
    let encounteredZero = false
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] == 0) {
            encounteredZero = true
            if (count + lastCount > longest) {
                longest = count + lastCount
            }
            lastCount = count
            count = 0
        } else {
            count++
        }

        if (i == nums.length - 1) {
            if (count + lastCount > longest) {
                longest = count + lastCount
            }
        }
    }
    if (!encounteredZero) {
        return longest - 1
    }
    return longest
}

function largestAltitude(gain: number[]): number {
    let highest = 0
    let current = 0
    for (let g of gain) {
        current += g
        if (current > highest) {
            highest = current
        }
    }
    return highest
}

function pivotIndex(nums: number[]): number {
    let prefixArr = new Array(nums.length)
    let suffixArr = new Array(nums.length)
    let prefix = 0
    let suffix = 0
    for (
        let i = 0, j = nums.length - 1;
        i < nums.length && j >= 0;
        i++, j--
    ) {
        prefixArr[i] = prefix
        suffixArr[j] = suffix
        prefix += nums[i]
        suffix += nums[j]
    }
    for (let i = 0; i < nums.length; i++) {
        if (prefixArr[i] == suffixArr[i]) {
            return i
        }
    }
    return -1
}

function findDifference(
    nums1: number[],
    nums2: number[]
): number[][] {
    const num1Set = new Set(nums1)
    const num2Set = new Set(nums2)
    for (let num of num2Set) {
        if (num in num1Set) {
            num1Set.delete(num)
            num2Set.delete(num)
        }
    }
    const num1 = Array.from(num1Set)
    const num2 = Array.from(num2Set)
    return [num1, num2]
}

function uniqueOccurrences(arr: number[]): boolean {
    let numObj = {}
    let occObj = {}
    for (let num of arr) {
        if (numObj[num]) {
            numObj[num]++
        } else {
            numObj[num] = 1
        }
    }
    for (let num in numObj) {
        if (occObj[numObj[num]]) {
            return false
        } else occObj[numObj[num]] = 1
    }
    return true
}

function removeStars(s: string): string {
    const answer: string[] = []
    for (let i = 0; i < s.length; i++) {
        if (s[i] === "*") {
            answer.pop()
            continue
        }
        answer.push(s[i])
    }
    return answer.join("")
}

function decodeString(s: string): string {
    const currentNumber: number[] = []
    let answer = ""

    function getString(i: number, number: string): any {
        currentNumber.push(Number(number))
        let strings = ""
        let newNumber = ""
        let string
        for (let j = i + 1; j < s.length; j++) {
            if (s[j] === "[") {
                ;[j, string] = getString(j, newNumber)
                newNumber = ""
                strings += string
                continue
            }
            if (s[j] === "]") {
                strings = strings.repeat(
                    currentNumber.pop() as number
                )
                return [j, strings]
            }
            if (Number(s[j]) || Number(s[j]) === 0) {
                newNumber += s[j]
            } else {
                strings += s[j]
            }
        }
    }

    let number = ""
    for (let i = 0; i < s.length; i++) {
        let strings = ""
        if (s[i] === "[") {
            ;[i, strings] = getString(i, number)
            number = ""
            answer += strings
            continue
        }
        if (Number(s[i]) || Number(s[i]) === 0) {
            number += s[i]
        } else {
            answer += s[i]
        }
    }
    return answer
}

function equalPairs(grid: number[][]): number {
    let num = 0
    let rows: any = {}
    let columns: string[] = []

    for (let i = 0; i < grid.length; i++) {
        let row = ""
        let column = ""
        for (let j = 0; j < grid.length; j++) {
            row += `-${grid[i][j]}`
            column += `-${grid[j][i]}`
        }
        if (rows[row]) {
            rows[row] += 1
        } else {
            rows[row] = 1
        }

        columns.push(column)
    }

    for (let i = 0; i < columns.length; i++) {
        if (rows[columns[i]]) {
            num += rows[columns[i]]
        }
    }
    return num
}

function asteroidCollision(asteroids: number[]): number[] {
    let output: number[] = []
    for (let i = 0; i < asteroids.length; i++) {
        if (asteroids[i] > 0) {
            output.push(asteroids[i])
            continue
        } else if (output.length === 0) {
            output.push(asteroids[i])
            continue
        } else {
            let still_crushing = true
            do {
                if (
                    output.length == 0 ||
                    output[output.length - 1] < 0
                ) {
                    output.push(asteroids[i])
                    still_crushing = false
                    continue
                }
                let collision =
                    output[output.length - 1] + asteroids[i]
                if (collision > 0) {
                    still_crushing = false
                } else if (collision == 0) {
                    output.pop()
                    still_crushing = false
                } else {
                    output.pop()
                }
            } while (still_crushing)
        }
    }
    return output
}

class RecentCounter {
    private pings: number[]
    constructor() {
        this.pings = []
    }

    ping(t: number): number {
        this.pings.push(t)
        let count = 0
        for (let i = this.pings.length - 1; i >= 0; i--) {
            if (this.pings[i] >= t - 3000) {
                count++
            } else {
                break
            }
        }
        // const count = this.pings.filter((ping) => {
        //     if (ping >= t - 3000) {
        //         return true
        //     }
        //     return false
        // })
        return count
    }
}
